### PR TITLE
Small performance improvements to sorting

### PR DESCRIPTION
**Pull request by mkeskells**
Reduce some allocation in sorting for small cases where we only have a couple of elements to sort, which seems to happen a lot

Also for List, we can avoid the builder. We could apply this pattern to other cases

### CODE REVIEW

**Comment by Ichoran** [comment]
Is this enough faster to be worth it?

If this is enough faster, should we also specialize other small cases (e.g. 3-5)?  Sorting networks generally are pretty fast for fixed numbers of items, and even a fixed-stack bubble/insertion sort isn't too bad, and you could have a series of methods like

```scala
sortFiveInto[A, CC...](b: Builder[...], a1: A, a2: A, a3: A, a4: A, a5: A)(implicit ord: Ordering[A]) {
  var x1 = a1
  ...
  var x5 = a5
  if (ord.<(x2, x1)) { val x = x1; x1 = x2; x2 = x }
  if (ord.<(x3, x1)) ...
  ...
  if (ord.<(x2, x3)) ...
  ...
  if (ord.<(x4, x5)) { val x = x4; x4 = x5; x5 = x }
  b += x1
  ...
  b += x5
  b.result
}
```

File: src/library/scala/collection/immutable/Seq.scala [hunk 1]
```diff
@@ -36,6 +37,20 @@ trait Seq[+A] extends Iterable[A]
   override def toSeq: Seq[A] = this
   override def seq: Seq[A] = this
   protected[this] override def parCombiner = ParSeq.newCombiner[A] // if `immutable.SeqLike` gets introduced, please move this there!
+  override def sorted[B >: A](implicit ord: Ordering[B]): this.type = {
```
Comment (mkeskells) [inline src/library/scala/collection/immutable/Seq.scala]: Not really sure it this is the correct place to insert the override

There isn't an `immutable.SeqLike` which would be the ideal I think

Comment (NthPortal) [inline src/library/scala/collection/immutable/Seq.scala]: I don't think `this.type` is a valid return type, as it's not returning `this`

File: src/library/scala/collection/immutable/Stream.scala [hunk 2]
```diff
@@ -201,6 +201,7 @@ import scala.language.implicitConversions
  *  @define orderDependentFold
  *  @define willTerminateInf Note: lazily evaluated; will terminate for infinite-sized collections.
  */
+@SerialVersionUID(4897901117128916054L)
```
Comment (dwijnand) [inline src/library/scala/collection/immutable/Stream.scala:204]: Was this added intentionally?  We have a few other PRs about serialisation, so I wouldn't want this flying in by mistake.

**Comment by dwijnand** [comment]
This looks like it need more attention, with individual commits failing as well as the end result.

**Comment by SethTisue** [comment]
closing — provisionally, as always! — for inactivity

**Pull request closed without merging**